pkg/controller/volume/attachdetach/util/util_test.go

```diff
@@ -0,0 +1,348 @@
+			},
+		},
+		{
+			desc:           "pv type that supports csi migration",
+			createNodeName: nodeName,
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "kube-apiserver",
+					Namespace: "default",
+				},
+				Spec: v1.PodSpec{
+					Volumes: []v1.Volume{
+						{
+							Name: migratedVolume,
+							VolumeSource: v1.VolumeSource{
+								PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+									ClaimName: "migrated-pvc",
+									ReadOnly:  false,
+								},
+							},
+						},
+					},
+				},
+			},
+			wantPersistentVolume: &v1.PersistentVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: migratedVolume,
+				},
+				Spec: v1.PersistentVolumeSpec{
+					PersistentVolumeSource: v1.PersistentVolumeSource{
+						CSI: &v1.CSIPersistentVolumeSource{
+							Driver:           "pd.csi.storage.gke.io",
+							VolumeHandle:     "projects/UNSPECIFIED/zones/UNSPECIFIED/disks/test-disk",
+							FSType:           "ext4",
+							ReadOnly:         false,
+							VolumeAttributes: map[string]string{"partition": ""},
+						},
+					},
+					ClaimRef: &migratedObjectReference,
+				},
+			},
+		},
+		{
+			desc:           "CSINode not found for a volume type that supports csi migration",
+			createNodeName: kubetypes.NodeName("another-node"),
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "kube-apiserver",
+					Namespace: "default",
+				},
+				Spec: v1.PodSpec{
+					Volumes: []v1.Volume{
+						{
+							Name: migratedVolume,
+							VolumeSource: v1.VolumeSource{
+								PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+									ClaimName: "migrated-pvc",
+									ReadOnly:  false,
+								},
+							},
+						},
+					},
+				},
+			},
+			wantErrorMessage: "csiNode \"another-node\" not found",
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			plugMgr, intreeToCSITranslator, csiTranslator, pvLister, pvcLister := setup(testNodeName, t)
+			actualSpec, err := CreateVolumeSpec(test.pod.Spec.Volumes[0], test.pod, test.createNodeName, plugMgr, pvcLister, pvLister, intreeToCSITranslator, csiTranslator)
+
+			if actualSpec == nil && (test.wantPersistentVolume != nil || test.wantVolume != nil) {
+				t.Errorf("got volume spec is nil")
+			}
+
+			if (len(test.wantErrorMessage) > 0 && err == nil) || (err != nil && !strings.Contains(err.Error(), test.wantErrorMessage)) {
+				t.Errorf("got err %v, want err with message %v", err, test.wantErrorMessage)
+			}
+
+			if test.wantPersistentVolume != nil {
+				gotVolumeWithCSIMigration := *actualSpec.PersistentVolume
+				if gotVolumeWithCSIMigration.Name != test.wantPersistentVolume.Name {
+					t.Errorf("got volume name is %v, want volume name is %v", gotVolumeWithCSIMigration.Name, test.wantPersistentVolume.Name)
+
+				}
+				if !reflect.DeepEqual(gotVolumeWithCSIMigration.Spec, test.wantPersistentVolume.Spec) {
+					if !reflect.DeepEqual(gotVolumeWithCSIMigration.Spec.CSI, test.wantPersistentVolume.Spec.CSI) {
```

Comment:
I still think this inner layer of checks is not needed.

You can use `cmp.Diff` make it easier to debug failures. 
```
                                if !reflect.DeepEqual(gotVolumeWithCSIMigration.Spec, test.wantPersistentVolume.Spec) {
-                                       if !reflect.DeepEqual(gotVolumeWithCSIMigration.Spec.CSI, test.wantPersistentVolume.Spec.CSI) {
-                                               t.Errorf("got volume.Spec.CSI %v, want.Spec.CSI %v", gotVolumeWithCSIMigration.Spec.CSI, test.wantPersistentVolume.Spec.CSI)
-                                       }
-                                       if gotVolumeWithCSIMigration.Spec.CSI == nil {
-                                               t.Errorf("got volume.Spec %v, want.Spec %v", gotVolumeWithCSIMigration, test.wantPersistentVolume)
-                                       }
+                                       t.Errorf("unexpected volume spec got/want diff\n%s", cmp.Diff(gotVolumeWithCSIMigration.Spec, test.wantPersistentVolume.Spec))
                                }
```

Here's what I found:
```
--- FAIL: Test_CreateVolumeSpec (0.00s)
    --- FAIL: Test_CreateVolumeSpec/inline_volume_type_that_supports_csi_migration (0.00s)
        util_test.go:258: unexpected volume spec got/want diff
              v1.PersistentVolumeSpec{
                Capacity:                      nil,
                PersistentVolumeSource:        {CSI: &{Driver: "pd.csi.storage.gke.io", VolumeHandle: "projects/UNSPECIFIED/zones/UNSPECIFIED/disks/test-disk", FSType: "ext4", VolumeAttributes: {"partition": ""}, ...}},
            -   AccessModes:                   []v1.PersistentVolumeAccessMode{"ReadWriteOnce"},
            +   AccessModes:                   nil,
                ClaimRef:                      nil,
                PersistentVolumeReclaimPolicy: "",
                StorageClassName:              "",
                MountOptions:                  nil,
            -   VolumeMode:                    &"Filesystem",
            +   VolumeMode:                    nil,
                NodeAffinity:                  nil,
              }
FAIL
```

Reply:
We need the inner check because deepEqual cannot tell the promoted structure. reflect.DeepEqual(gotVolumeWithCSIMigration.Spec, test.wantPersistentVolume.Spec) is false when the inner structure is gotVolumeWithCSIMigration.Spec.CSI and test.wantPersistentVolume.Spec.PersistentVolumeSource.CSI even when it is equal.